server/src/main/java/org/elasticsearch/index/codec/vectors/IVFVectorsReader.java

```diff
@@ -0,0 +1,354 @@
+    public final void checkIntegrity() throws IOException {
+        rawVectorsReader.checkIntegrity();
+        CodecUtil.checksumEntireFile(ivfCentroids);
+        CodecUtil.checksumEntireFile(ivfClusters);
+    }
+
+    @Override
+    public final FloatVectorValues getFloatVectorValues(String field) throws IOException {
+        return rawVectorsReader.getFloatVectorValues(field);
+    }
+
+    @Override
+    public final ByteVectorValues getByteVectorValues(String field) throws IOException {
+        return rawVectorsReader.getByteVectorValues(field);
+    }
+
+    protected float[] getGlobalCentroid(FieldInfo info) {
+        if (info == null || info.getVectorEncoding().equals(VectorEncoding.BYTE)) {
+            return null;
+        }
+        FieldEntry entry = fields.get(info.number);
+        if (entry == null) {
+            return null;
+        }
+        return entry.globalCentroid();
+    }
+
+    @Override
+    public final void search(String field, float[] target, KnnCollector knnCollector, Bits acceptDocs) throws IOException {
+        final FieldInfo fieldInfo = state.fieldInfos.fieldInfo(field);
+        if (fieldInfo.getVectorEncoding().equals(VectorEncoding.FLOAT32) == false) {
+            rawVectorsReader.search(field, target, knnCollector, acceptDocs);
+            return;
+        }
+        // TODO add new ivf search strategy
+        int nProbe = 10;
+        float percentFiltered = 1f;
+        if (acceptDocs instanceof BitSet bitSet) {
+            percentFiltered = Math.max(0f, Math.min(1f, (float) bitSet.approximateCardinality() / bitSet.length()));
+        }
+        int numVectors = rawVectorsReader.getFloatVectorValues(field).size();
+        BitSet visitedDocs = new FixedBitSet(state.segmentInfo.maxDoc() + 1);
+        IntPredicate needsScoring = docId -> {
+            if (acceptDocs != null && acceptDocs.get(docId) == false) {
+                return false;
+            }
+            return visitedDocs.getAndSet(docId) == false;
+        };
+
+        FieldEntry entry = fields.get(fieldInfo.number);
+        CentroidQueryScorer centroidQueryScorer = getCentroidScorer(
+            fieldInfo,
+            entry.postingListOffsets.length,
+            entry.centroidSlice(ivfCentroids),
+            target,
+            ivfClusters
+        );
+        final NeighborQueue centroidQueue = scorePostingLists(fieldInfo, knnCollector, centroidQueryScorer, nProbe);
+        PostingVisitor scorer = getPostingVisitor(fieldInfo, ivfClusters, target, needsScoring);
+        int centroidsVisited = 0;
+        long expectedDocs = 0;
+        long actualDocs = 0;
+        // initially we visit only the "centroids to search"
+        while (centroidQueue.size() > 0 && centroidsVisited < nProbe) {
+            ++centroidsVisited;
+            // todo do we actually need to know the score???
+            int centroidOrdinal = centroidQueue.pop();
+            // todo do we need direct access to the raw centroid???
```

Comment:
+1 for this TODO: it looks like we always do 2 reads of quantized and full centroid, why we need only a single

Reply:
We need the raw to quantize the query vector for the posting list. So, we score with the quantized and then quantize the query vector according to the raw value. 

I think we might want to separate out the scoring vs. the iterating for the centroids.